Add Sedimentation Tutorial Langevin vs. LB

pkreissl opened the pull request:
Two-dimensional lattice-Boltzmann sedimentation tutorial to show the impact of hydrodynamics effects on particle diffusion.

pkreissl committed:
Add draft for sedimentation tutorial.

jngrad commented on doc/tutorials/lattice_boltzmann/lattice_boltzmann_sedimentation.ipynb:
```diff
@@ -0,0 +1,396 @@
+    "data_flowfield = np.full((sampling_steps, n_height, n_width, 3), np.nan)\n",
+    "\n",
+    "# return n_x x n_x array of lattice velocities, averaged in z-direction \n",
+    "def mean_flowfield(xs=n_width, ys=n_height, zs=n_depth):\n",
+    "    dataframe = np.array([[[lbf[x, y, z].velocity for z in range(zs)] for x in range(xs)] \n",
+    "                          for y in range(ys)])\n",
+    "    return np.mean(dataframe, axis=2)\n",
```
Consider using a LB profile observable here, a particle position observable and a particle velocity observable, and accumulators. This way we don't need to re-implement core features as numpy operations, and can remove the numpy arrays pre-filled with NaN values.

jngrad commented on doc/tutorials/lattice_boltzmann/lattice_boltzmann_sedimentation.ipynb:
```diff
@@ -0,0 +1,396 @@
+    "    ax.hlines((1, box_height-1), 0, box_width, color=\"gray\")\n",
+    "\n",
+    "# create meshgrid for quiver plot\n",
+    "xs = np.array([x for x in range(n_width)]) * spacing\n",
+    "ys = np.array([y for y in range(n_height)]) * spacing\n",
+    "X, Y = np.meshgrid(xs, ys)\n",
+    "\n",
+    "# initialize plot objects\n",
+    "lb_ff = ax2.quiver(X, Y, data_flowfield[0, :, :, 0],\n",
+    "                   data_flowfield[0, :, :, 1])\n",
+    "lb_particles, = ax2.plot([], [], 'o')\n",
+    "lv_particles, = ax1.plot([], [], 'o')\n",
+    "\n",
+    "def draw_frame(t):\n",
+    "    # manually remove Quivers from ax2\n",
+    "    for artist in ax2.get_children():\n",
+    "        if isinstance(artist, Quiver):\n",
+    "            artist.remove()\n",
+    "    \n",
+    "    # draw new quivers\n",
+    "    lb_ff = ax2.quiver(X, Y, data_flowfield[t, :, :, 0],\n",
+    "                       data_flowfield[t, :, :, 1])\n",
```
By default, matplotlib quivers are automatically scaled. As a result, at the beginning of the simulation when the LB velocity is in the range $10^{-12}$, the arrows flip up and down to adjust to the random noise, which causes the screen to flicker. Also, the fluid velocity inside the boundary is used to create arrows that give the impression the top LB boundary creates a fluid flow. Please find below a solution that rescales arrow lengths based on the largest velocity in the entire trajectory and removes the top and bottom rows of the LB fluid.

```suggestion
    "xs = np.array([x for x in range(n_width)]) * spacing\n",
    "ys = np.array([y for y in range(1, n_height-1)]) * spacing\n",
    "X, Y = np.meshgrid(xs, ys)\n",
    "\n",
    "# set quiver scale (fraction of the highest velocity in the XY plane)\n",
    "lb_vel_max = np.sum(np.square(data_flowfield[:, :, :, 0:-1]), axis=-1)\n",
    "quiver_scale = np.max(lb_vel_max) / 4.\n",
    "\n",
    "def plot_lb_vel(ax, X, Y, data_flowfield, t, scale):\n",
    "    # skip data inside the LB boundaries (top and bottom walls)\n",
    "    return ax.quiver(X, Y,\n",
    "                     data_flowfield[t, 1:-1, :, 0],\n",
    "                     data_flowfield[t, 1:-1, :, 1],\n",
    "                     scale_units=\"xy\", scale=scale)\n",
    "\n",
    "# initialize plot objects\n",
    "lb_ff = plot_lb_vel(ax2, X, Y, data_flowfield, 0, quiver_scale)\n",
    "lb_particles, = ax2.plot([], [], 'o')\n",
    "lv_particles, = ax1.plot([], [], 'o')\n",
    "\n",
    "def draw_frame(t):\n",
    "    # manually remove Quivers from ax2\n",
    "    for artist in ax2.get_children():\n",
    "        if isinstance(artist, Quiver):\n",
    "            artist.remove()\n",
    "    \n",
    "    # draw new quivers\n",
    "    lb_ff = plot_lb_vel(ax2, X, Y, data_flowfield, t, quiver_scale)\n",
```

jngrad reviewed (commented):
Overall LGTM. The CPU and GPU version produce the same flow. The waLBerla version however takes longer to reach the break of symmetry, at which point the particles are already 90% down the channel.

jngrad committed:
doc: Finish LB sedimentation tutorial

Use normalized quiver plots. Speed-up sampling using observables.
Introduce crystalline defects to the sides of the box to observe
the drop at the center of the box. Explain LD and LB thermostats.

jngrad committed:
tests: Check LB sedimentation tutorial

kodiakhq[bot] committed:
Merge branch 'python' into lb_sedimentation_tutorial
